src/iop/censorize.c

```diff
@@ -0,0 +1,456 @@
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  darktable is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include "bauhaus/bauhaus.h"
+#include "common/debug.h"
+#include "common/gaussian.h"
+#include "common/opencl.h"
+#include "control/control.h"
+#include "develop/develop.h"
+#include "develop/imageop.h"
+#include "develop/imageop_math.h"
+#include "develop/imageop_gui.h"
+#include "develop/noise_generator.h"
+#include "develop/tiling.h"
+#include "gui/accelerators.h"
+#include "gui/gtk.h"
+#include "gui/presets.h"
+#include "iop/iop_api.h"
+#include <assert.h>
+#include <gtk/gtk.h>
+#include <math.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <inttypes.h>
+
+#define CLAMPF(a, mn, mx) ((a) < (mn) ? (mn) : ((a) > (mx) ? (mx) : (a)))
+
+DT_MODULE_INTROSPECTION(1, dt_iop_censorize_params_t)
+
+typedef struct dt_iop_censorize_params_t
+{
+  float radius_1;              // $MIN: 0.0 $MAX: 500.0 $DEFAULT: 0.0  $DESCRIPTION: "input blur radius"
+  float pixelate;              // $MIN: 0.0 $MAX: 500.0 $DEFAULT: 0.0 $DESCRIPTION: "pixellation radius"
+  float radius_2;              // $MIN: 0.0 $MAX: 500.0 $DEFAULT: 0.0  $DESCRIPTION: "output blur radius"
+  float noise;                 // $MIN: 0.0 $MAX: 1.0   $DEFAULT: 0.0   $DESCRIPTION: "noise level"
+} dt_iop_censorize_params_t;
+
+
+typedef struct dt_iop_censorize_gui_data_t
+{
+  GtkWidget *radius_1, *pixelate, *radius_2, *noise;
+} dt_iop_censorize_gui_data_t;
+
+typedef dt_iop_censorize_params_t dt_iop_censorize_data_t;
+
+typedef struct dt_iop_censorize_global_data_t
+{
+  int kernel_lowpass_mix;
+} dt_iop_censorize_global_data_t;
+
+typedef struct point_t
+{
+  size_t x, y;
+} point_t;
+
+const char *
+name()
+{
+  return _("censorize");
+}
+
+const char *description(struct dt_iop_module_t *self)
+{
+  return dt_iop_set_description(self, _("censorize license plates and body parts for privacy"),
+                                      _("creative"),
+                                      _("linear or non-linear, RGB, scene-referred"),
+                                      _("frequential, RGB"),
+                                      _("special, RGB, scene-referred"));
+}
+
+int flags()
+{
+  return IOP_FLAGS_INCLUDE_IN_STYLES | IOP_FLAGS_SUPPORTS_BLENDING | IOP_FLAGS_ALLOW_TILING;
+}
+
+int default_group()
+{
+  return IOP_GROUP_EFFECT | IOP_GROUP_EFFECTS;
+}
+
+int default_colorspace(dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe, dt_dev_pixelpipe_iop_t *piece)
+{
+  return iop_cs_rgb;
+}
+
+#if FALSE
```

Comment:
Why ? Is that debug code ? If really deactivated is that because process_cl() is not working ? So either:
- remove the routine
- fix it
- or add comment saying why it is comment out

Reply:
The OpenCL path is not implemented, but I let the template function there because it only needs minor changes to work (the whole module is built upon a copy of the lowpass module, so the gaussian blur kernels can be reused from there and only the noise and pixellation kernels need to be written).

Reply:
Comment added